src/Sub.re

```diff
@@ -0,0 +1 @@
+include Sub_internal;
```

Comment:
Why this?

Reply:
Oh, I think I see, because internals are used in the `SubscriptionRunner`. My philosophy on this is that all top-level modules are internal, and that the external interface should be out only in the main public module. So in `Isolienar.rei` you'd instead have:

```
...
module Sub: {
  ...
}
```

Reply:
Right, this is the issue - I use the internals in `SubscriptionRunner`.

I like your idea a lot! I tried it out here: https://github.com/bryphe/isolinear/pull/10

Unfortunately, I can't get the types to compile, though. The issue is that the test case uses `SubscriptionRunner`, which expects the 'internal' `Sub.t`, and it doesn't know that this is compatible with the 'external'-facing `Sub.t`.

That manifests as error messages like this: 
```
       but an expression was expected of type testState Isolinear__Sub.t
File "test/StoreTests.re", line 44, characters 19-242:
44 | ...................{
45 |           type msg = bool;
46 |           type model = bool;
47 | 
48 |           let initial = false;
49 |           let updater = (_model, msg) => (msg, Effect.none);
50 |           let subscriptions = model => model ? sub : Sub.none;
51 |         }..
Error: Signature mismatch:
       ...
       Values do not match:
         val subscriptions : model -> model Sub.t
       is not included in
         val subscriptions : model -> model Isolinear__Sub.t
       File "src/Store.rei", line 35, characters 6-44: Expected declaration
       File "test/StoreTests.re", line 50, characters 14-27:
```

Let me know if you have ideas, though!

Reply:
Why does it expect the internal one?

Reply:
I added a note in the PR here: https://github.com/bryphe/isolinear/pull/10

>This is an attempt (that doesn't work) to remove the ugly type equality in Sub.rei:

`type t('msg) = Sub_internal.t('msg)`

This is needed because we use the SubscriptionRunner in tests, and without that equality, the SubscriptionRunner doesn't know that the `Isolinear__Sub.t` created through the public interface is the same as the internal `Sub.t` that it uses.

Not sure how to fix this, but maybe you have ideas?

Reply:
Adding proper interfaces to the exports might help surface the conflict, or at least make it easier to understand. But it looks like it's at least partly because you expose internals for testing. So why do you need that? The interface should be purely functional and completely observable using the public interface. If it isn't, that's a bit of a red flag in itself I think.

Reply:
I do like the idea of making exposed internals namespaced though. But i think it should be at the functional level, and would prefer the module to be named `Testing` instead of `Internal`, to more strongly suggest that it's only to be used for testing. So instead of `Isolinear.Internal.Sub.whatever` I'd like to have `Isolinear.Sub.Testing.whatever`. That might also make it easier to use the same types for the public and internal interfaces.

Reply:
> So why do you need that?

`SubscriptionRunner` isn't exposed publicly - it's an implementation detail of the `Store`. But I do wish to test it. That's the root of the issue.

You could argue that it should purely be tested via the public API, but I do think there is value in testing `SubscriptionRunner` itself, independently - failures there are more scoped and easier to diagnose than failures at a higher-level.

So we really have a trade-off between two options:
Option 1) Only test the public API, avoid this issue entirely
Option 2) Add unit test for internals, but with the ugliness of the equality

The best-case-mythical-option-3 would be to unit test the internals w/o needing the type equality - but I'm not sure there is a way? So we're stuck between deciding what the best option is, given the trade-offs, between option 1 and option 2. 

If I understand correctly, I think where we might differ in opinion is which set of trade-offs is the right one to make - you'd rather lean towards option 1, and we have option 2 in place now. 

> So instead of Isolinear.Internal.Sub.whatever I'd like to have Isolinear.Sub.Testing.whatever. 

I don't have a strong preference on the naming - `Internal` vs `Testing` - so I'd be happy to switch to whatever is more intuitive.

> That might also make it easier to use the same types for the public and internal interfaces.

Hmm, I don't understand how it would help with this issue - could you clarify this please? Maybe this could solve our problem and not require us to make trade-offs between option 1 and option 2 above.

Reply:
> If I understand correctly, I think where we might differ in opinion is which set of trade-offs is the right one to make 

No, not necessarily. I mostly just forgot that `SubscriptionRunner` was internal 😄 

> Hmm, I don't understand how it would help with this issue - could you clarify this please?

If you define the public interface of `SubscriptionRunner` in the same file as the abstract `Sub.t`, you can that type instead of the internal one for the functions in `SubscriptionRunner`. Then hopefully the compiler will either be able to figure out the type equality, or at least give a better indication of where the conflict is.

Let me have a closer look at it though, to see if I can figure it out.